### ISSUE TITLE
useLastNotificationResponse clears only when app is killed.

### ISSUE THREAD

**Issue by kryptowiz** [expo/expo#18712]
### Summary

We are using the `useLastNotificationResponse` hook to check against tapped notifications and route the user to a specific screen. 

**However, we can't figure out how to clear the data from that hook at all. The last notification data only clears out when the app is killed.**

In our case, our users are now always being routed until they kill the app. 

We can figure out a hacky way where we save the last notification and do comparisons against incoming ones but that's obviously not ideal.

### What platform(s) does this occur on?

Android, iOS

### Environment

```
  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 12.2.1
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 16.15.0 - /usr/local/bin/node
      Yarn: 1.22.19 - ~/Documents/work/mobile/node_modules/.bin/yarn
      npm: 8.5.5 - /usr/local/bin/npm
    Managers:
      CocoaPods: 1.11.3 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
    IDEs:
      Xcode: 13.4/13F17a - /usr/bin/xcodebuild
    npmPackages:
      expo: ^45.0.0 => 45.0.6 
      react: 17.0.2 => 17.0.2 
      react-dom: 17.0.2 => 17.0.2 
      react-native: 0.68.2 => 0.68.2 
      react-native-web: 0.17.7 => 0.17.7 
    npmGlobalPackages:
      eas-cli: 0.53.1
      expo-cli: 6.0.2
    Expo Workflow: managed
```

### Minimal reproducible example

```
const lastNotification = Notifications.useLastNotificationResponse();

const handleBackgroundNotification = async () => {
    const route = lastNotification?.notification.request.content.data.route;			
    routeUser(data);
    await clearNotifications();
};

useEffect(() => {
    handleBackgroundNotification();
}, [lastNotification]);
```

**Comment by mlalitthapa** [comment]
Experiencing the same issue.

**Comment by douglowder** [comment]
Reopening this and will take a look. Probably the right approach is to add a `clearLastNotificationResponseAsync()` API for SDK 52. @vonovak

**Comment by ggat** [comment]
Hi @douglowder,

`Notifications.clearLastNotificationResponseAsync()`, doesn't seem to affect `Notifications.addNotificationResponseReceivedListener()`. 

If you clear the notification response, unsubscribe and re-subscribe with `addNotificationResponseReceivedListener()` you get the last response as if `clearLastNotificationResponseAsync()` didn't work

**Comment by douglowder** [comment]
@ggat that is very interesting, I will try to repro. Are you seeing this on Android, iOS, or both?

**Comment by ggat** [comment]
@douglowder it's on  iOS 18.1.1, expo 52.0,  expo-notifications 0.29. I didn't have a chance to test it on Android 🙏

**Comment by douglowder** [comment]
@ggat I'm having trouble reproducing the issue with last notification response coming back when re-adding a listener. Could you share the code you are using to test this?

**Comment by mrchappie** [comment]
hey @douglowder I am experiencing the same issues and this is my code snapshot and the problem is that after I interact with a notification, every time I reload the development build, last response is not null nor undefined and it is the info of the last notification I have interacted with.
    ```
useEffect(() => {
              Notifications.setNotificationHandler({
                handleNotification: async () => ({
                  shouldShowAlert: true,
                  shouldPlaySound: true,
                  shouldSetBadge: true,
                }),
              });
                      // CHECK PENDING RESPONSE
                      async function checkPendingResponse() {
                        const lastResponse = await Notifications.getLastNotificationResponseAsync();
                        // console.log(lastResponse, '   pend res');
                        if (lastResponse) {
                          const actionId = lastResponse.actionIdentifier;
                          handleNotificationResponse(
                            actionId,
                            lastResponse.notification.request.identifier,
                            lastResponse.notification.request.content.data.todo as TodoI,
                            {
                              screen: lastResponse.notification.request.content.data.screen,
                              params: lastResponse.notification.request.content.data.params,
                            } as ScreenToNavigateI
                          );
                        }
                      }
                  
                      checkPendingResponse();
    }, []);
```

**Comment by appsgenie** [comment]
I am experiencing the exact same issue when calling `clearLastNotificationResponseAsync` after reading and acting on the notification from `getLastNotificationResponseAsync`. reopening the app just keeps taking me back to my screen that is according to the notification. I currently have a fix to track this inside async storage but there are other features/actions that clear the storage and it is not an ideal solution... so far I was able to reproduce this on iOS only. And yes reopening the app remounts all components which is where the `addNotificationResponseReceivedListener` is added.

**Comment by dustinvforsyth** [comment]
So I finally figured out what my issue was with this and from the code @mrchappie I am pretty sure he had the same setup.

The issue I think that is throwing everyone off is anytime a notification launches the app it looks like `expo-notifications` is saving that as the last response. I thought it was only saving the last response if the notification wasn't handled. Which honestly doesn't make any sense now thinking through it because the native code wouldn't know if you handled it or not.

So the fix for me was that I needed to call `Notifications.clearLastNotificationResponseAsync()` in both handlers:

```
      Notifications.getLastNotificationResponseAsync().then((notification) => {
        Notifications.clearLastNotificationResponseAsync()
        onNotificationReceived(notification);
      });
      const listener = Notifications.addNotificationResponseReceivedListener((notification) => {
        Notifications.clearLastNotificationResponseAsync()       <<<<<< -----------------------------I was missing this one.
        onNotificationReceived(notification);
      });

```
Honestly ended up adding it to my `onNotificationRecieved()` function in the end since that was the common function between the two but this just shows the issue better. 

Hope that helps someone else!
